stubs/fanstatic/fanstatic/compiler.pyi

```diff
@@ -0,0 +1,121 @@
+from _typeshed import StrOrBytesPath
+from abc import abstractmethod
+from logging import Logger
+from subprocess import Popen
+from typing import Any, ClassVar
+from typing_extensions import Literal
+
+import setuptools.command.sdist
+from fanstatic.core import Resource
+
+logger: Logger
+
+class CompilerError(Exception): ...
+
+class Compiler:
+    @property
+    @abstractmethod
+    def name(self) -> str: ...
+    @property
+    @abstractmethod
+    def source_extension(self) -> str: ...
+    def __call__(self, resource: Resource, force: bool = False) -> None: ...
+    def process(self, source: StrOrBytesPath, target: StrOrBytesPath) -> Any: ...
+    def should_process(self, source: StrOrBytesPath, target: StrOrBytesPath) -> bool: ...
+    @property
+    @abstractmethod
+    def available(self) -> bool: ...
+    def source_path(self, resource: Resource) -> str | None: ...
+    def target_path(self, resource: Resource) -> str | None: ...
+
+class Minifier(Compiler):
+    @property
+    @abstractmethod
+    def name(self) -> str: ...
+    @property
+    @abstractmethod
+    def source_extension(self) -> str: ...
+    @property
+    @abstractmethod
+    def target_extension(self) -> str: ...
+    def source_to_target(self, resource: Resource) -> str: ...
+
+def compile_resources(argv: list[str] = ...) -> None: ...
+
+class sdist_compile(setuptools.command.sdist.sdist): ...
+
+class NullCompiler(Compiler):
+    name: ClassVar[Literal[""]]
+    source_extension = NotImplemented
+    def source_path(self, resource: Resource) -> None: ...
+    def target_path(self, resource: Resource) -> None: ...
+    def should_process(self, source: StrOrBytesPath, target: StrOrBytesPath) -> Literal[False]: ...
+    @property
+    def available(self) -> Literal[False]: ...
+
+SOURCE: object
+TARGET: object
+
+class CommandlineBase:
+    @property
+    @abstractmethod
+    def command(self) -> str: ...
+    arguments: ClassVar[list[str]]
+    @property
+    def available(self) -> bool: ...
+    def process(self, source: StrOrBytesPath, target: StrOrBytesPath) -> Popen[str]: ...
```

Comment:
`SOURCE` and `TARGET` can be used as sentinel arguments. `_typeshed/__init__.pyi` has this recipe for sentinel values:

```
_SourceType = NewType("_SourceType", object)
_TargetType = NewType("_TargetType", object)
SOURCE: _SourceType
TARGET: _TargetType
# [...]
def process(self, source: StrOrBytesPath | _SourceType, target: StrOrBytesPath | _TargetType) -> ...
```